File: src/main/java/ru/octol1ttle/flightassistant/hud/impl/AlertDisplay.java
package ru.octol1ttle.flightassistant.hud.impl;

import net.minecraft.client.font.TextRenderer;
import net.minecraft.client.gui.DrawContext;
import net.minecraft.text.Text;
import ru.octol1ttle.flightassistant.Dimensions;
import ru.octol1ttle.flightassistant.DrawHelper;
import ru.octol1ttle.flightassistant.alerts.api.BaseAlert;
import ru.octol1ttle.flightassistant.alerts.api.ICenteredAlert;
import ru.octol1ttle.flightassistant.alerts.api.IECAMAlert;
import ru.octol1ttle.flightassistant.computers.impl.TimeComputer;
import ru.octol1ttle.flightassistant.computers.impl.safety.AlertController;
import ru.octol1ttle.flightassistant.config.FAConfig;
import ru.octol1ttle.flightassistant.hud.api.IHudDisplay;
import ru.octol1ttle.flightassistant.registries.ComputerRegistry;

public class AlertDisplay implements IHudDisplay {
    private final Dimensions dim;
    private final AlertController alert = ComputerRegistry.resolve(AlertController.class);
    private final TimeComputer time = ComputerRegistry.resolve(TimeComputer.class);

    public AlertDisplay(Dimensions dim) {
        this.dim = dim;
    }

    @Override
    public void render(DrawContext context, TextRenderer textRenderer) {
        if (!FAConfig.indicator().showAlerts) {
            return;
        }
        if (ComputerRegistry.isFaulted(AlertController.class)) {
            renderFaulted(context, textRenderer, Text.translatable("alerts.flightassistant.fault.computers.alert_mgr"));
            return;
        }
        boolean renderedCentered = false;
        int x = dim.lFrame + 5;
        int y = dim.tFrame + 15;

        for (BaseAlert alert : alert.activeAlerts) {
            if (!renderedCentered && alert instanceof ICenteredAlert centered) {
                renderedCentered = centered.render(textRenderer, context, dim.xMid,
                        dim.yMid + 10, time.highlight);
            }

            if (!alert.hidden && alert instanceof IECAMAlert ecam) {
                y += 10 * ecam.render(textRenderer, context, x, y, time.highlight);
            }
        }
    }

    @Override
    public void renderFaulted(DrawContext context, TextRenderer textRenderer) {
        renderFaulted(context, textRenderer, Text.translatable("alerts.flightassistant.fault.indicators.alert"));
    }

    private void renderFaulted(DrawContext context, TextRenderer textRenderer, Text text) {
        DrawHelper.drawHighlightedText(textRenderer, context, text, dim.lFrame + 5, dim.tFrame + 15,
                FAConfig.indicator().warningColor, time.highlight);
    }

}

File: src/client/kotlin/ru/octol1ttle/flightassistant/impl/display/AlertDisplay.kt
package ru.octol1ttle.flightassistant.impl.display

import net.minecraft.client.gui.DrawContext
import net.minecraft.text.*
import net.minecraft.util.Identifier
import ru.octol1ttle.flightassistant.FlightAssistant
import ru.octol1ttle.flightassistant.api.alert.*
import ru.octol1ttle.flightassistant.api.computer.ComputerAccess
import ru.octol1ttle.flightassistant.api.display.*
import ru.octol1ttle.flightassistant.api.util.*
import ru.octol1ttle.flightassistant.config.FAConfig

class AlertDisplay : Display() {
    private val withUnderline: Style = Style.EMPTY.withUnderline(true)

    override fun enabled(): Boolean {
        return FAConfig.display.showAlerts
    }

    override fun render(drawContext: DrawContext, computers: ComputerAccess) {
        with(drawContext) {
            val x: Int = HudFrame.left + 10
            var y: Int = HudFrame.top + 5

            var renderedCentered = false
            for (category: AlertCategory in computers.alert.categories) {
                val copy: MutableText = category.categoryText.copy()
                drawText(
                    copy.setStyle(withUnderline),
                    x,
                    y,
                    (category.getHighestPriority() ?: continue).colorSupplier.invoke()
                )
                copy.append(" ")

                for (alert: Alert in category.activeAlerts) {
                    if (!renderedCentered && alert is CenteredAlert) {
                        alert.render(this, computers, centerYI + 8)
                        renderedCentered = true
                        y += fontHeight
                    }

                    if (alert is ECAMAlert) {
                        y += fontHeight * alert.render(this, computers, x + getTextWidth(copy), x, y)
                    }
                }
                y += 3
            }
        }
    }

    override fun renderFaulted(drawContext: DrawContext) {
        with(drawContext) {
            val x: Int = HudFrame.left + 10
            val y: Int = HudFrame.top + 5

            drawText(Text.translatable("short.flightassistant.alert"), x, y, warningColor)
        }
    }

    companion object {
        val ID: Identifier = FlightAssistant.id("alert")
    }
}
